File: src/main/java/org/droidmate/analyzer/AppUnderTest.java
package org.droidmate.analyzer;

import net.dongliu.apk.parser.ApkFile;
import net.dongliu.apk.parser.bean.ApkMeta;
import org.apache.commons.io.FileUtils;
import org.droidmate.analyzer.api.DummyApkMeta;
import org.droidmate.analyzer.api.IApi;
import org.droidmate.analyzer.exploration.ExplorationResult;
import org.droidmate.analyzer.exploration.IExplorationStrategy;
import org.droidmate.analyzer.exploration.IScenario;
import org.droidmate.analyzer.exploration.Scenario;
import org.droidmate.analyzer.wrappers.BoxMateWrapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Application under evaluation
 */
public class AppUnderTest {
    private static final Logger logger = LoggerFactory.getLogger(AppUnderTest.class);

    private Configuration cfg;
    private BoxMateWrapper boxMate;
    private ApkFile apk;
    private Path apkFile;
    private List<IScenario> scenarios;
    private int currExplDepth;
    private Path dir;

    AppUnderTest(Configuration cfg, Path path) {
        this.scenarios = new ArrayList<>();
        this.currExplDepth = 0;
        this.cfg = cfg;
        this.boxMate = new BoxMateWrapper(this.cfg);
        this.apkFile = path.toAbsolutePath();

        try {
            this.apk = new ApkFile(this.apkFile.toFile());
        } catch (IOException e) {
            logger.error(e.getMessage(), e);
        }

        assert this.apk != null;
        this.createExperimentDir();
    }

    private void createExperimentDir() {
        ApkMeta meta = this.getMeta();
        String dirName = String.format("%s_%s", meta.getPackageName(), meta.getVersionName());
        this.dir = Paths.get(this.cfg.dataDir.toString(), dirName);

        try {
            Files.createDirectories(this.dir);
            FileUtils.cleanDirectory(this.dir.toFile());
        } catch (IOException e) {
            logger.error(e.getMessage(), e);
        }

        assert Files.exists(this.dir);
    }

    public Path getApkFile() {
        return this.apkFile;
    }

    private ApkMeta getMeta() {
        ApkMeta apkMeta = null;
        try {
            apkMeta = this.apk.getApkMeta();
        } catch (IOException e) {
            logger.error(e.getMessage(), e);
        }

        if (apkMeta == null)
            return new DummyApkMeta();

        return apkMeta;
    }

    public Path getDir() {
        return this.dir;
    }

    public IScenario getInitialExpl() {
        Optional<IScenario> scenarioStream = this.scenarios.stream()
                .filter(p -> p.getExplDepth() == 0)
                .findFirst();

        return scenarioStream.orElse(null);
    }

    public int getCurrExplDepth() {
        return this.currExplDepth;
    }

    public List<IScenario> getScenarios(int depth) {
        return this.scenarios.stream()
                .filter(p -> p.getExplDepth() == depth)
                .collect(Collectors.toList());
    }

    private void addScenarios(Collection<IScenario> expl) {
        assert (expl != null);

        if (expl.size() > 0)
            this.scenarios.addAll(expl);
    }

    private List<IScenario> getPendingScenarios() {
        return this.scenarios.stream()
                .filter(p -> p.getResult() == null)
                .collect(Collectors.toList());
    }

    private boolean hasPendingScenarios() {
        return this.getPendingScenarios().size() > 0;
    }

    public String getPackageName() {
        return this.getMeta().getPackageName();
    }

    private void inline(IScenario scenario) {
        // Inline app
        Path inlinedFile = boxMate.inlineApp(this.getApkFile(), scenario.getCfgFile());
        scenario.setInlinedApk(inlinedFile);
        assert scenario.getInlinedApk() != null;
    }

    private List<IApi> getInitialApiList(){
        IScenario initialExpl = this.getInitialExpl();
        if (initialExpl == null)
            return new ArrayList<>();

        return initialExpl.getExploredApiList();
    }

    public List<IApi> getInitialMonitoredApiList(){
        return this.getInitialApiList().stream()
                .filter(IApi::hasRestriction)
                .distinct()
                .collect(Collectors.toList());
    }

    void explore(IExplorationStrategy strategy) {
        // Initial expl
        List<IScenario> initialExpl = strategy.generateScenarios(this);
        this.addScenarios(initialExpl);

        while (this.hasPendingScenarios()) {
            for (IScenario scenario : this.getPendingScenarios()) {
                this.inline(scenario);
                ExplorationResult explRes;

                explRes = boxMate.explore(scenario.getInlinedApk(), scenario.getExplDepth() == 0);
                scenario.setResult(explRes);
                assert (scenario.getResult() != null) && (Files.exists(scenario.getResult().getExplDir()));
            }

            ++this.currExplDepth;
            List<IScenario> newScenarios = strategy.generateScenarios(this);
            this.addScenarios(newScenarios);
        }
    }
}

File: src/main/kotlin/org/droidmate/analyzer/AppUnderTest.kt
package org.droidmate.analyzer

import net.dongliu.apk.parser.ApkFile
import net.dongliu.apk.parser.bean.ApkMeta
import org.apache.commons.io.FileUtils
import org.droidmate.analyzer.api.DummyApkMeta
import org.droidmate.analyzer.api.IApi
import org.droidmate.analyzer.exploration.IExplorationStrategy
import org.droidmate.analyzer.exploration.IScenario
import org.droidmate.analyzer.wrappers.BoxMateWrapper
import org.droidmate.analyzer.wrappers.IBoxMateWrapper
import org.slf4j.LoggerFactory
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.util.*

class AppUnderTest internal constructor(private val cfg: Configuration, override val apkFile: Path,
                                        private val boxMate: IBoxMateWrapper = BoxMateWrapper(cfg)) : IAppUnderTest {
    private var apk: ApkFile? = null
    override val scenarios: MutableList<IScenario> = ArrayList()
    override var currExplDepth: Int = 0
        private set
    override var dir: Path = Constants.EMPTY_PATH
        private set

    init {
        this.currExplDepth = 0

        try {
            this.apk = ApkFile(this.apkFile.toFile())
        } catch (e: IOException) {
            logger.error(e.message, e)
        }

        assert(this.apk != null)
        this.dir = this.createExperimentDir()
    }

    private fun createExperimentDir(): Path {
        val meta = this.meta
        val dirName = String.format("%s_%s", meta.packageName, meta.versionName)
        val newDir = this.cfg.dataDir.resolve(dirName)

        try {
            Files.createDirectories(newDir)
            FileUtils.cleanDirectory(newDir.toFile())
        } catch (e: IOException) {
            logger.error(e.message, e)
        }

        assert(Files.exists(newDir))

        return newDir
    }

    private val meta: ApkMeta
        get() {
            var apkMeta: ApkMeta? = null
            try {
                apkMeta = this.apk!!.apkMeta
            } catch (e: IOException) {
                logger.error(e.message, e)
            }

            if (apkMeta == null)
                return DummyApkMeta()

            return apkMeta
        }

    override val initialExpl: IScenario?
        get() {
            val scenarioStream = this.scenarios.stream()
                    .filter { p -> p.explDepth == 0 }
                    .findFirst()

            return scenarioStream.orElse(null)
        }

    override fun getScenariosDepth(depth: Int): List<IScenario> {
        return this.scenarios
                .filter { p -> p.explDepth == depth }
                .toList()
    }

    private fun addScenarios(expl: Collection<IScenario>) {
        if (expl.isNotEmpty())
            this.scenarios.addAll(expl)
    }

    private val pendingScenarios: List<IScenario>
        get() = this.scenarios
                .filter { p -> p.result == null }
                .toList()

    private fun hasPendingScenarios(): Boolean {
        return this.pendingScenarios.isNotEmpty()
    }

    override val packageName: String
        get() = this.meta.packageName

    private fun inline(scenario: IScenario) {
        // Inline app
        if ((this.initialExpl != null) && (this.initialExpl!!.inlinedApk != Constants.EMPTY_PATH))
            scenario.inlinedApk = this.initialExpl!!.inlinedApk
        else
            scenario.inlinedApk = boxMate.inlineApp(this.apkFile)
    }

    private val initialApiList: List<IApi>
        get() {
            return this.initialExpl!!.exploredApiList
        }

    override val initialMonitoredApiList: List<IApi>
        get() = this.initialApiList
                .filter { p -> p.hasRestriction() }
                .distinct()
                .toList()

    override fun explore(strategy: IExplorationStrategy) {
        // Initial expl
        val initialExpl = strategy.generateScenarios(this)
        this.addScenarios(initialExpl)

        while (this.hasPendingScenarios()) {
            for (scenario in this.pendingScenarios) {
                this.inline(scenario)
                scenario.result = boxMate.explore(scenario.inlinedApk, scenario.cfgFile, scenario.explDepth == 0)
                assert(scenario.result != null && Files.exists(scenario.result!!.explDir))
            }

            ++this.currExplDepth
            val newScenarios = strategy.generateScenarios(this)
            this.addScenarios(newScenarios)
        }
    }

    override fun toString(): String {
        return this.apkFile.fileName.toString()
    }

    override val successfulScenarios: List<IScenario>
        get() = this.scenarios
                .filter { p -> p.isValid }
                .toList()

    override val failScenarios: List<IScenario>
        get() = this.scenarios
                .filterNot { p -> p.isValid }
                .toList()

    companion object {
        private val logger = LoggerFactory.getLogger(AppUnderTest::class.java)
    }
}
